An editor mode for adding and updating bookmarks

ZwodahS opened the pull request:
by specifying --editor, this will open up an editor to edit, similar to how git commit message works.

This will be useful for long descriptions.
Currently only support adding

jarun commented:
:+1: Simply awesome!!! I'll review and pull it in. Please allow me some time.

Questions (haven't seen the code yet):

1. Is it possible to use a custom editor?
2. What's the default editor.
3. Are you willing to extend this feature to update as well?
4. Can you please provide some more details on the implementation in the first comment?

jarun commented:
Also, we need to document new options in man, README, add an example (this would go quite high because using an editor is very convenient).

ZwodahS commented:
1. The editor that is used can be configured using the $EDITOR env var.
2. Default to vim
3. Will definitely extend this to update, but I will need some time to find the best place to put it in. Perhaps someone more familiar with the codebase can help.
4. The implementation is similar to that in http://stackoverflow.com/questions/6309587/call-up-an-editor-vim-from-a-python-script. I have wrapped it around a function which can be used to open up the editor with existing values.

I will put it in the man and README later today. Do you prefer a single squashed commit or okay with multiple commits ?

jarun commented:
> Default to vim

We support Windows too. We either need to (a) suppress this option on Windows (don't even show) or (b) fail saying editor not found (applies to all platforms). I prefer (b).

What if both manual option and editor is specified? e.g.:

    buku -a xyz.com --comment help help --editor

> Perhaps someone more familiar with the codebase can help.

I will surely help you. Please let me know the questions. I think you can initially invoke the editor and pass the parsed data to `update_rec()` when option `--update ID --editor` is specified.

> The implementation is similar to

This really helps someone touching the code after a few months. I keep forgetting stuff myself. ;)

> I will put it in the man and README later today.

Thank you!

> Do you prefer a single squashed commit or okay with multiple commits ?

A single commit would be great! Again because it helps anyone referring to the changes later.

ZwodahS commented:
yeah.. windows would be tricky. Some users might have vim/emacs installed and they should be allowed to use this too. Another way is to disable this unless the user set the EDITOR env var, instead of defaulting to vim.

> buku -a xyz.com --comment help help --editor
In this case, the editor will be shown with "help help" populated in the section for comment.
Similarly, if you specified tags in the commandline args, it will also appear in the editor.

> I will surely help you. Please let me know the questions. I think you can initially invoke the editor and pass the parsed data to update_rec() when options --update ID --editor are specified together.

Yeap this certainty seems like the way that i should do it. 

I will take a look and integrate the update first, then squashed them and submit another PR. Let me know how you feel about only enabling it for users with $EDITOR set.

ZwodahS commented:
I have added the --editor mode for updating existing bookmark. It will only work if you specify a single bookmark to update, and will fail otherwise.

Also, I have disable --editor by default unless $EDITOR is set in the env var. Let me know if you want this to be changed.

I also updated the README and the help menu. 

Let me know if i missed anything

jarun commented:
> Another way is to disable this unless the user set the EDITOR env var

How about making `--editor` accept an optional arg:
- if no args passed to `--editor`, look for `$EDITOR` (and fail if not defined)
- else, try to execute the arg, fail if execution fails
?

For a quick test, I ran:

    $ ./buku.py -a www.google.com --editor
and got the following screen:

![](https://i.imgur.com/geDy8WS.png)

- Can we put the instructions above? e.g.
`insert title **below** this line, in a single line`
etc.?
- We can use directives like
`# TITLE goes below this line (in a single line)`
the CAPS make it easier to read.
- The following should be 2 lines:
`# insert the tags above in a single line, comma separated# insert all the comments/descriptions below`
- The `# ----- #` should have a blank line above it, if this is the end marker.

> if you specified tags in the commandline args, it will also appear in the editor.

If the title is *not* specified and not added in editor, it should be fetched from the web. This is broken right now.

jarun commented:
> Also, I have disable --editor by default unless $EDITOR is set in the env var.

I think you pushed and I commented almost at the same time :). Please check my notes on `--update` in my last comment.

From my comment:

> If the title is not specified and not added in editor, it should be fetched from the web. This is broken right now.

I missed the case where user doesn't even want to add a title. What if we instruct and treat just a `-` in the editor (for the comment) as the marker of *truly blank title*? If the line is just blank we fetch the title from the web as usual.

> It will only work if you specify a single bookmark to update, and will fail otherwise.

We are good as long as it's documented in program help and main.

One the workflow is finalized, we can add a subsection on `$EDITOR` under `Operational notes` in README and `ENVIRONMENT` in man page.

ZwodahS commented:
No worries, I also like the idea of getting it right without having to revisit it again.

1. Will update the instructional messages to be more clearer.

> I missed the case where user doesn't even want to add a title. What if we instruct and treat just a - in the editor (for the comment) as the marker of truly blank title? If the line is just blank we fetch the title from the web as usual.

2. So for title, empty line will means to fetch from web, and a single '-' for empty title. I totally miss the part for fetching from web because I was working offline and didn't test that part.

> else, try to execute the arg, fail if execution fails

3. I would rather fail if the editor is not found, than to try resolving it as much as possible, since passing --editor is a explicit intention. 

Lastly, can i hijack '-x' for single character param ? Typing --editor is really long-winded

jarun commented:
> Will update the instructional messages to be more clearer.

Thank you!

> So for title, empty line will means...

Exactly!

> since passing --editor is a explicit intention.

Agreed.

> Will update --editor to accept the editor name

Thanks!

> can i hijack '-x' for single character param

Absolutely. Would `-w` (for write, writer) be more memorable?

jarun commented:
Also, I think I can squash the commits together while merging. So you don't need to raise a new PR. We have all the relevant discussion here.

ZwodahS commented:
Fixed various stuffs mentioned above. will be using '-w' as suggested.

jarun commented:
1. I do not have any editor named heli ;). So:

```
$ ./buku.py -a www.google.com/ --editor heli
Traceback (most recent call last):
  File "./buku.py", line 2838, in <module>
    main()
  File "./buku.py", line 2566, in main
    result = open_editor(args.editor, url, title_in, tags, desc_in)
  File "./buku.py", line 2208, in open_editor
    subprocess.call([editor, temp.name])
  File "/usr/lib/python3.5/subprocess.py", line 557, in call
    with Popen(*popenargs, **kwargs) as p:
  File "/usr/lib/python3.5/subprocess.py", line 947, in __init__
    restore_signals, start_new_session)
  File "/usr/lib/python3.5/subprocess.py", line 1551, in _execute_child
    raise child_exception_type(errno_num, err_msg)
FileNotFoundError: [Errno 2] No such file or directory: 'heli'
```

2. In editor:

```
www.google.com/                                                                               
# insert LINK **above** this line. (single line)                                
                                                                                
# insert TITLE **above** this line, (single line) (empty line to auto fetch, "-" for empty title)
,                                                                               
# insert the TAGS **above** this line, comma separated. (single line)           
hello                                                                           
# insert all COMMENTS **above** this line. (multiple lines)
```

can we make this:

```
# LINK goes below this line (single line).
www.google.com/                                                                               
# TITLE goes below this line (single line). Leave empty line to auto fetch, "-" for empty title.
,                                                                               
# TAGS (comma-separated) go below this line (single line).           
hello                                                                           
# COMMENTS go below this line (multiple lines).

```

jarun commented on buku.py:
```diff
@@ -2182,6 +2184,107 @@ def sigint_handler(signum, frame):
+    # TITLE
+    if title_in == '':
+        title_in = '-'
+    elif title_in is None:
```
I think this check should go first.

jarun commented on buku.py:
```diff
@@ -2182,6 +2184,107 @@ def sigint_handler(signum, frame):
+    elif title == '-':
+        title = ''
+
+    tags = ","
```
'single quotes, single quotes, single all the way.'

jarun commented on buku.py:
```diff
@@ -2182,6 +2184,107 @@ def sigint_handler(signum, frame):
+    return "\n".join(strings)
+
+def parse_temp_file_content(content):
+    # refactor this to return error code instead ?
```
please remove these source comments.

jarun commented on README.md:
```diff
@@ -459,11 +460,34 @@ The same number of iterations must be specified for one lock & unlock instance.
-30. More **help**:
+30. **Open In editor**
+
+        $ buku --editor -a https://ddg.gg search engine, privacy
```
`--write`?

jarun commented on README.md:
```diff
@@ -459,11 +460,34 @@ The same number of iterations must be specified for one lock & unlock instance.
+
+## Editor mode file format.
+
+`$ buku --editor -a https://ddg.gg search engine, privacy`
```
`--write`?

jarun commented on buku.py:
```diff
@@ -2269,6 +2372,9 @@ def main():
                          -a: do not set title, -u: clear title
     -c, --comment [...]  description of the bookmark, works with
                          -a, -u; clears comment, if no arguments
+    -w, --write [editor] open editor to write instead of args.
```
Check earlier comment on description.

jarun commented on buku.py:
```diff
@@ -2277,6 +2383,7 @@ def main():
     addarg('--tag', nargs='*', help=HIDE)
     addarg('-t', '--title', nargs='*', help=HIDE)
     addarg('-c', '--comment', nargs='*', help=HIDE)
+    addarg('-w', '--editor', nargs='*', help=HIDE)
```
`--write`?

jarun commented on buku.py:
```diff
@@ -2454,7 +2561,14 @@ def main():
             tags = parse_tags(keywords[1:])
 
-        bdb.add_rec(args.add[0], title_in, tags, desc_in, args.immutable)
+        url = args.add[0]
```
Can we also ensure `buku --write` opens the editor (even if `-a` is not passed) and adds a bookmark if added from editor?

ZwodahS replied:
sounds like a good idea, let me add that

jarun replied:
:+1:

jarun commented on buku.py:
```diff
@@ -2454,7 +2561,14 @@ def main():
-        bdb.add_rec(args.add[0], title_in, tags, desc_in, args.immutable)
+        url = args.add[0]
+        if args.editor is not None:
+            result = open_editor(args.editor, url, title_in, tags, desc_in)
```
Probably we can have a separate API to check editor is available or not and call that from add and update option processing from inside main().

jarun commented on buku.py:
```diff
@@ -2545,31 +2659,52 @@ def main():
-                        lower, upper = upper, lower
+            if args.editor is not None: # check for editor mode
+                # currently allow only editing of one url
+                if len(args.update) != 1 or not is_int(args.update[0]):
```
Limit the number of args using `nargs='?'` in addarg(). Not so deep down below.

ZwodahS replied:
This is tricky. How do know if -w is present ?

jarun replied:
Same way you do now:

```
if args.write is not None:` # it is not passed
    if args.write: # it is not an empty list
```

ZwodahS replied:
when i use nargs='?', args.write is both None for when i specify -w without argument and when I don't specify -w.

jarun replied:
Oops!

ZwodahS replied:
I am going to use '*', to allow us the flexibility to specify the editor in $EDITOR or -w [editor], will that do ?

jarun replied:
`'--foo', nargs='?', const='no_args'`

jarun commented:
I've just completed the code review. Please address those as well. I'm hoping we can have this feature in in a few hours.

ZwodahS commented on buku.py:
```diff
@@ -2182,6 +2184,107 @@ def sigint_handler(signum, frame):
+        temp.write(temp_file_content)
+        temp.flush()
+        subprocess.call([editor, temp.name])
+        with open(temp.name) as f: # for some reason sometimes this don't get saved properly
```
Sorry, i think this wasn't clear.

The problem is that after you save your editors, sometimes the temp object doesn't get updated with the new content. The workaround I did here is to open the temp file again and read it, instead of using the temp object.

jarun replied:
Ahh OK. Then make sure the line is < 80 chars, otherwise move the comment above the line.

ZwodahS commented on auto-completion/bash/buku-completion.bash:
```diff
@@ -16,6 +16,7 @@ _buku () {
         -d --delete
         --deep
         -e --export
+        --editor
```
Hmm I am not sure If i should blindly add them in, since I can't really test them. 
will update it with -w and --write for bash completion

jarun replied:
OK. I'll add those.

ZwodahS commented:
The above cleans up the code and README. Allowing to edit without specifying -a and -u will be in another commit.

jarun commented:
I've made several modifications on the base commit. Please review d1c1164e16ea299d7b8692bdb92de5e4e3c7787f.

Also, I tried the GUI editor mousepad. I couldn't update a record. It gets aborted. is there any way to handle output from GUI editors?

ZwodahS commented:
It looks like you clean up the args parsing quite nicely. There is one behaviour that is changed.
When the user specify '-w' without any args and no $EDITOR is set, there is no error message. Previously a error message is displayed informing the users that they need to set the EDITOR or pass the arg. If you are okay with this then is fine 😄 

As for GUI editor, I tried it with macvim. For macvim, when you run it in the command line, it will open up in the background.
However by specifying '-f' it will changed it to opening macvim in foreground with our temp file.

`subprocess.call([editor, '-f', temp.name])`

My guess is that every editor have a flag like this. What we need is to allow the user to specify something like this in the commandline or EDITOR.

A quick fix for this is to 
```
command = editor.split(' ')
command.append(temp.name)
subprocess.call(command)
```

This will allow users to specify the 'arg' to be passed in as well. I will see if there are other ways tomorrow.

jarun commented:
> There is one behaviour that is changed.
I need to fix this. Thanks!

I believe mousepad forks internally and returns 0 from parent. Can't be handled. We don't care.

```
>>> import subprocess
>>> subprocess.check_call(["mousepad", "hello"])
0
>>>
```

> it will open up in the background

Opening in the foreground or background is not the problem. `Buku` exits immediately after calling mousepad. Are you seeing the same with macvim?

jarun commented:
> There is one behaviour that is changed.

Fixed now!

ZwodahS commented:
by default macvim opens in the background and exits immediately.

However, specifying '-f' opens it and buku will wait for macvim to finish before continuing.

jarun commented:
How about:

`buku -w 'macvim -f'`

Does that work?

ZwodahS commented:
It will works with the code i pasted above that split editor arg into a
list.

I think the sub process command don't work with arg with spaces
On Sun, 5 Feb 2017 at 2:26 PM, Arun Prakash Jana <notifications@github.com>
wrote:

> How about:
>
> buku -w 'macvim -f'
>
> Does that work?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jarun/Buku/pull/112#issuecomment-277500406>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAMC-y_mtVokAivLNtZwsVejDsjIlxqGks5rZWubgaJpZM4L0Hzd>
> .
>

jarun commented:
I'll add it then. Thanks!

jarun commented:
Fixed in issue #115.

jarun commented:
The idea was fantastic, and `buku -w` by itself is a priceless contribution. I'm loving it myself now! ;) Thanks a ton!

ZwodahS commented:
haha thank you too. I was going to build something myself, specifically the -w part, but realized that it will be easier if i just add it in to existing one.

Cheers
